modules/core/src/encoder/field/PurposeRestrictionVectorEncoder.ts

```diff
@@ -21,57 +21,74 @@ export class PurposeRestrictionVectorEncoder {
         bitString += IntEncoder.encode(purpRestriction.purposeId, BitLength.purposeId);
         bitString += IntEncoder.encode(purpRestriction.restrictionType, BitLength.restrictionType);
 
-        // now get all the vendors under that restriction
+        // Get available vendor IDs from the GVL
+        const gvlVendors = [...prVector.gvl.vendorIds];
+
+        // Get all vendors to which the restriction applies
         const vendors: number[] = prVector.getVendors(purpRestriction);
-        const len: number = vendors.length;
+
+        // Counts number of element occurances in the array
+        const countElementOccurrences = (array, item): number =>
+          array.reduce(
+            (result, currentValue) => (currentValue === item ? result + 1 : result),
+            0,
+          );
 
         /**
-         * numEntries comes first so we will have to keep a counter and the do
-         * the encoding at the end
+         * Generate vendor ranges.
+         *
+         * When generating vendor range we are looking
+         * for the consecutive vendor IDs in the vendor list
          */
-        let numEntries = 0;
-        let startId = 0;
-        let rangeField = '';
+        const filteredGvlVendors = gvlVendors.map((currentItem, index) =>
+          // Mark unused vendors from the vendor list with null
+          vendors.includes(currentItem) ? index : null,
```

Comment:
I'm a bit concerned with the `O[n*M]` performance here since we're using `Array.includes` in a long loop. But it's probably okay, but if there's something easy that would be a direct lookup it would be even better.

Reply:
This has got a lotta loops over potentially a big list of vendors.  First destructuring the vendorIds `Set` into an `Array`, then map x includes, then filter, then reduce.  

[If you change this `if` within the for loop](https://github.com/InteractiveAdvertisingBureau/iabtcf-es/blob/2e45df478387e27a0b93487f2ea1482e519bc0d5/modules/core/src/encoder/field/PurposeRestrictionVectorEncoder.ts#L48)
```typescript
// either end of the loop or there's a gap greater than 1 number
if (i === len - 1 || vendors[i + 1] > vendorId + 1) {
```
from "close the range if it's the end of the list or the next restricted vendor is not the sequential number"

```typescript
/**
 * either end of the loop or there's a gap greater than 1 number and
 * the vendorId is on the GVL
 * /
 if (i === len - 1 || (vendors[i + 1] > vendorId + 1 && prVector.gvl.vendorIds.has(vendorId + 1))) {
```
to "close the range if it's the end of the list or the next restricted vendor is not the sequential number **_AND_** that number isn't used."

ie. check the `Set` of vendorIds on the GVL (O(1)) lookup to see if it's not used, and therefore

Reply:
That's the only change needed to this entire class...

Reply:
Agreed and updated the logic as suggested by @chrispaterson